docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1056 @@
+
+```cs
+unsafe int UnmanagedPointersSum(Span<int> buffer)
+{
+    fixed (int* pBuffer = buffer)
+    {
+        int* pEnd = pBuffer + buffer.Length;
+        int* pOneVectorFromEnd = pEnd - Vector128<int>.Count;
+        int* pCurrent = pBuffer;
+
+        Vector128<int> sum = Vector128<int>.Zero;
+
+        while (pCurrent <= pOneVectorFromEnd)
+        {
+            sum += Vector128.Load(pCurrent);
+
+            pCurrent += Vector128<int>.Count;
+        }
+
+        int result = Vector128.Sum(sum);
+
+        while (pCurrent < pEnd)
+        {
+            result += *pCurrent;
+
+            pCurrent++;
+        }
+
+        return result;
+    }
+}
+```
+
+`LoadAligned` and `LoadAlignedNonTemporal` require the input to be aligned. Aligned reads and writes should be slightly faster but using them comes at a price of increased complexity.
+
+Currently .NET exposes only one API fo allocating unmanaged aligned memory: [NativeMemory.AlignedAlloc](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativememory.alignedalloc). In the future, we might provide [a dedicated API](https://github.com/dotnet/runtime/issues/27146) for allocating managed, aligned and hence pinned memory buffers.
+
+The alternative to creating aligned buffers (we don't always have the control over input) is to pin the buffer, find first aligned address, handle non-aligned elements, then start aligned loop and afterwards handle the remainder. Adding such complexity to our code is hardly ever worth it and needs to be proved with proper benchmarking on various hardware.
+
+The fourth method expects only a managed reference (`ref T source`). We don't need to pin the buffer (GC is tracking managed references and updates them if memory gets moved), but it still requires us to properly handle managed pointer arithmetic:
+
+```cs
+int ManagedReferencesSum(int[] buffer)
+{
+    ref int current = ref MemoryMarshal.GetArrayDataReference(buffer);
+    ref int end = ref Unsafe.Add(ref current, buffer.Length);
+    ref int oneVectorAwayFromEnd = ref Unsafe.Subtract(ref end, Vector128<int>.Count);
+
+    Vector128<int> sum = Vector128<int>.Zero;
+
+    while (!Unsafe.IsAddressGreaterThan(ref current, ref oneVectorAwayFromEnd))
+    {
+        sum += Vector128.LoadUnsafe(ref current);
+
+        current = ref Unsafe.Add(ref current, Vector128<int>.Count);
+    }
+
+    int result = Vector128.Sum(sum);
+
+    while (Unsafe.IsAddressLessThan(ref current, ref end))
+    {
+        result += current;
+
+        current = ref Unsafe.Add(ref current, 1);
+    }
+
+    return result;
+}
+```
+
+**Note:** `Unsafe` does not expose a method called "IsGreaterOrEqualThan", so we are using a negation of `Unsafe.IsAddressGreaterThan` to achieve desired effect.
```

Comment:
> `Unsafe` does not expose a method called "IsGreaterOrEqualThan"

I think it'd be clearer to say "does not expose a method call "IsLessThanOrEqualTo"", as those are the semantics of the loop in question and also the negation of `IsAddressGreaterThan`.

Reply:
Great catch!